backend/src/test/java/fk/prof/backend/LeaderAPILoadAndAssociationTest.java

```diff
@@ -0,0 +1,221 @@
+
+@RunWith(VertxUnitRunner.class)
+public class LeaderAPILoadAndAssociationTest {
+  private Vertx vertx;
+  private Integer port;
+  private DeploymentOptions leaderHttpDeploymentOptions;
+
+  private TestingServer testingServer;
+  private CuratorFramework curatorClient;
+  private List<Recorder.ProcessGroup> mockProcessGroups;
+
+
+  @Before
+  public void setBefore(TestContext context) throws Exception {
+    mockProcessGroups = Arrays.asList(
+      Recorder.ProcessGroup.newBuilder().setAppId("a").setCluster("c").setProcName("p1").build(),
+      Recorder.ProcessGroup.newBuilder().setAppId("a").setCluster("c").setProcName("p2").build(),
+      Recorder.ProcessGroup.newBuilder().setAppId("a").setCluster("c").setProcName("p3").build()
+    );
+    ConfigManager.setDefaultSystemProperties();
+
+    testingServer = new TestingServer();
+    curatorClient = CuratorFrameworkFactory.newClient(testingServer.getConnectString(), 500, 500, new RetryOneTime(1));
+    curatorClient.start();
+    curatorClient.blockUntilConnected(10, TimeUnit.SECONDS);
+
+    JsonObject config = ConfigManager.loadFileAsJson(LeaderAPILoadAndAssociationTest.class.getClassLoader().getResource("config.json").getFile());
+    JsonObject vertxConfig = ConfigManager.getVertxConfig(config);
+    vertx = vertxConfig != null ? Vertx.vertx(new VertxOptions(vertxConfig)) : Vertx.vertx();
+
+    JsonObject httpServerConfig = ConfigManager.getLeaderHttpServerConfig(config);
+    assert httpServerConfig != null;
+    port = httpServerConfig.getInteger("port");
+
+    JsonObject leaderHttpConfig = ConfigManager.getLeaderHttpDeploymentConfig(config);
+    assert leaderHttpConfig != null;
+    String backendAssociationPath = leaderHttpConfig.getString("backend.association.path");
+    curatorClient.create().forPath(backendAssociationPath);
+
+    leaderHttpDeploymentOptions = new DeploymentOptions(leaderHttpConfig);
+    BackendAssociationStore backendAssociationStore = VertxManager.getDefaultBackendAssociationStore(
+        vertx,
+        curatorClient,
+        backendAssociationPath,
+        ConfigManager.getLoadReportIntervalInSeconds(config),
+        leaderHttpConfig.getInteger("allowed.report.skips")
+    );
+
+    VertxManager.deployLeaderHttpVerticles(vertx, httpServerConfig, leaderHttpDeploymentOptions, backendAssociationStore);
+  }
+
+  @After
+  public void tearDown(TestContext context) throws IOException {
+    System.out.println("Tearing down");
+    VertxManager.close(vertx).setHandler(result -> {
+      System.out.println("Vertx shutdown");
+      curatorClient.close();
+      try {
+        testingServer.close();
+      } catch (IOException ex) {
+      }
+      if (result.failed()) {
+        context.fail(result.cause());
+      }
+    });
+  }
+
+  @Test(timeout = 5000)
+  public void reportNewBackendLoad(TestContext context) {
+    makeRequestReportLoad(BackendDTO.LoadReportRequest.newBuilder().setIp("1").setLoad(0.5f).build())
```

Comment:
Same assert-code-path-was-taken kind of verification as mentioned above is required here too.

If ar.succeeded() returns false, we'll bypass the assert.

Also, if the callback is dispatched on a different thread, the test won't fail despite asserts failing.

Its better to capture the variables and assert inline, make sure they are dispatched on the same thread as the test method itself.